app/controllers/materials_controller.rb

```diff
@@ -0,0 +1,221 @@
+require 'rubygems'
+require 'zip'
+
+class MaterialsController < ApplicationController
+
+  before_action :set_and_authorize_event
+
+  # POST /events/1/upload_material
+  def upload_material
+    Dir.mkdir(@event.material_path) unless Dir.exists?(@event.material_path)
+
+    path = params[:path].to_s
+    file = params[:file_upload]
+
+    if !is_file?(file)
+      redirect_to event_path(@event), alert: t("events.material_area.no_file_given") and return
+    elsif invalid_pathname?(path) || invalid_filename?(file.original_filename)
+      redirect_to event_path(@event), alert: t("events.material_area.invalid_path_given") and return
+    elsif !File.directory?(File.join(@event.material_path, path))
+      redirect_to event_path(@event), alert: t("events.material_area.download_file_not_found") and return
+    end
+
+    file_path = File.join(@event.material_path, path, file.original_filename)
+    begin
+      File.write(file_path, file.read, mode: "wb")
+    rescue IOError
+      redirect_to event_path(@event), alert: t("events.material_area.saving_fails") and return
+    end
+
+    redirect_to event_path(@event), notice: I18n.t("events.material_area.success_message")
+  end
+
+  # POST /events/1/download_material
+  def download_material
+    file_full_path = File.join(@event.material_path, params[:file].to_s)
+
+    if !params.has_key?(:file)
```

Comment:
I moved the assignment to `file_full_path` up so that I could write the `if..elsif..` construct to check for errors without interruptions, I thought it made the code clearer. I agree that it looks kind of weird to assign the value before checking whether it's valid, but since `find_full_path` is only used after this check, it doesn't change anything to the way the method works.